### PR TITLE
improve connection protocol, more consistent connection handling

### PR DESCRIPTION
**Pull request by TimoGlastra**
There is a lot of inconsistency between methods signatures in connection and issue credential. Currently with the present proof protocol I'm trying to be more consistent, also updating the other protocols. For example the issue credential module methods doesn't return anything at all. This means it is impossible to know the id of the credential record that is just created.

This is because of an API limitation. If we call `createXXX` in the service we expect to get the message back. But sometimes we also need the record associated with it. We could store the message inside the record so we only need to return the record, but this imposes a real limitation on how we must store things.  We are now required to store all messages, because we need to return the record and the message. In the present proof protocol I found it is easier to just store the attachments instead of whole messages. So instead of the `credential-offer` message, we just store the offer attachment. This also prevents us from needing to decode the attachment every time we need to access it.

I get this can be a controversial change, so I'm open to other options, but the API is too limited at the moment and I want to fix this before adding present proof.

I already briefly discussed with @karimStekelenburg  and this is the approach we came up with:

```ts
// Current approach
 const { message, connectionRecord } = await this.connectionService.createInvitation();
```

There are however a few other approaches and I don't have a strong opinion:

```ts
// Return the message name, instead of just "message" for all message types.
 const { invitationMessage, connectionRecord } = await this.connectionService.createInvitation();
 const { requestMessage, connectionRecord } = await this.connectionService.createInvitation();

// More generic
 const { message, record } = await this.connectionService.createInvitation();

// With names
 const { message: invitation, record: connectionRecord } = await this.connectionService.createInvitation();

// Return tuple instead of named parameters. This allows to name the return values yourself
 const [invitation, connectionRecord] = await this.connectionService.createInvitation();
```

Some other changes from this PR:
- updated the connection tests, added a few more tests, and added a few more assertions to the connection service it self.
- update the signature of connection service methods. They now returned `createOutboundMessage` where in the credential service it just returns the message. The credentials module creates the outbound message and sends it. I think this is a better approach. The service just handler the creation of the message and the updating of the record. The module concerns about sending.

Signed-off-by: Timo Glastra <timo@animo.id>

### CODE REVIEW

**Comment by TimoGlastra** [comment]
@jakubkoci this is now updated according to what we discussed in the WG call

**Comment by TimoGlastra** [comment]
Fixed merge conflicts

**Review by jakubkoci — COMMENTED** [review]
I like those refactorings. I added just one point related to passing an invitation via public API. We don't have to address in this PR, but it would be good to align our thinking about that ;)

File: src/lib/modules/ConnectionsModule.ts [hunk 1]
```diff
@@ -67,15 +66,13 @@ export class ConnectionsModule {
    * @returns new connection record
    */
   public async receiveInvitation(
-    invitationJson: Record<string, unknown>,
+    invitation: ConnectionInvitationMessage,
```
[the remark this answers is not in the document]
Comment (TimoGlastra) [reply src/lib/modules/ConnectionsModule.ts:69]: Well from my usage of the API I've found paths:

1. Scan QR / Receive URL and want to directly accept it. Use `receiveInvitationFromUrl`
2. Scan QR / Receive URL and want to extract some information from it before calling `receiveInvitation`.
    - `decodeInvitationFromUrl` return `ConnectionInvitationMessage` so you'd need to do `message.toJSON` again before passing it to `receiveInvitation`.
    - Or you should decode the url yourself by extracting `?c_i=` parameter and decoding. Why not use `decodeInvitationFromUrl`?

[the remark this answers is not in the document]
Comment (jakubkoci) [reply src/lib/modules/ConnectionsModule.ts:69]: I see. I thought there could be a case when you get an invitation in JSON not in URL. But if you should get URL every time we don't have a problem :) It seems that https://github.com/hyperledger/aries-rfcs/blob/master/features/0434-outofband/README.md also specifies that message should be encoded in URL.

**Review by jakubkoci — APPROVED** [review]

**Pull request merged**